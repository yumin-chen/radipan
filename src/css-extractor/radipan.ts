import { ComponentType, ReactNode } from "react";
import { css, cva, cx } from "../cli/get-design-system";
import { outdir } from "../cli/get-config";
import {
  RecipeDefinition,
  RecipeVariantRecord,
} from "@radipan-design-system/types/recipe";
import { SystemStyleObject } from "@radipan-design-system/types";
import { appendFileSync, existsSync, mkdirSync, writeFileSync } from "fs";
import { RecipeProps, CssProps, Creatable } from "../core/radipan.d";
import { transpile } from "./transpiler/transpiler";

const EXPORT_FOLDER = `node_modules/${outdir}/exported`;
const process = (typeof global !== "undefined" && global?.process) || {
  env: { DEBUG: false, CSSGEN: "", CSSGEN_FILE: "" },
};
const DEBUG = process?.env?.DEBUG;

const _h = (component, props, children) => ({ component, props, children });

const getVariantProps = (props: RecipeProps) => {
  const { css: cssProp, ...restProps } = props;
  return Object.keys(cssProp?.variants || {}).reduce(
    (prev, variantName) => ({
      ...prev,
      [variantName]: restProps?.[variantName as keyof typeof restProps],
    }),
    {}
  );
};

export const parseCssProp = (props: CssProps) => {
  const { css: cssProp } = props;
  const exportFile = `${EXPORT_FOLDER}/${process.env.CSSGEN_FILE}.css.js`;
  DEBUG && console.debug("Writing to file:", exportFile);
  if (process.env.CSSGEN === "pregen" && !!process.env.CSSGEN_FILE) {
    const fileDirIndex = process.env.CSSGEN_FILE.lastIndexOf("/");
    const fileDir = process.env.CSSGEN_FILE.substring(0, fileDirIndex);
    const exportFileDir = `${EXPORT_FOLDER}/${fileDir}`;
    if (!existsSync(exportFileDir)) {
      mkdirSync(exportFileDir, { recursive: true });
      DEBUG && console.debug("Created folder:", exportFileDir);
    }
    if (!existsSync(exportFile)) {
      writeFileSync(
        exportFile,
        '"use strict";\nimport {css, cva, cx} from "../css"\n\n'
      );
      DEBUG && console.debug("Created file:", exportFile);
    }
  }
  // Recipes
  const isRecipe = Object.hasOwn(cssProp || {}, "variants");
  if (isRecipe) {
    const variantProps = getVariantProps(props);
    appendFileSync(
      exportFile,
      `cva(${JSON.stringify(cssProp)})(${JSON.stringify(variantProps)});\n`
    );
    DEBUG && console.debug("Generated a `cva` function in ", exportFile);
    return cva(cssProp as RecipeDefinition<RecipeVariantRecord>)(variantProps);
  } else {
    appendFileSync(exportFile, `css(${JSON.stringify(cssProp)});\n`);
    DEBUG && console.debug("Generated a `css` function in ", exportFile);
    return css(cssProp as SystemStyleObject);
  }
};

export async function createElement(
  component: string | ComponentType,
  props: Readonly<Record<string, any>> | undefined,
  ...children: JSX.Element[] | ReactNode[]
) {
  // @ts-ignore
  const kids = children.every(item => !item) ? undefined : children;

  // Exhaustively instantiate components for CSS extraction
  if (typeof component === "function") {
    // @ts-ignore
    await component({ ...props, children: kids });
  } else {
    DEBUG && console.debug("Non-function component, skipping: ", component);
  }

  if (typeof props?.css === "object") {
    const { css: cssProp, className, _source, ...restProps } = props;
    const otherProps = { ...restProps };

    DEBUG &&
      console.debug(
        "Found css prop used within comopnent: ",
        component,
        props.css,
        _source
      );

    const cssClasses = parseCssProp(props);
    Object.keys(cssProp?.variants || []).forEach(
      variantName => delete otherProps[variantName]
    );

    if (process.env?.CSSGEN === "pregen") {
      if (props.radipanId) {
        await transpile(props.radipanId, className, cssClasses);
      } else {
        DEBUG && console.error("radipanId is missing: ", component, props.css);
      }
    }

    return _h(
      component,
      {
        ...otherProps,
        // Merge class names with generated styles
        className: !className ? cssClasses : cx(cssClasses, className),
      },
      kids
    );
  } else {
    DEBUG &&
      console.debug(
        "No css prop found used with comopnent: ",
        component?.name || component?.displayName || component,
        props
      );
  }

  return _h(component, props, kids);
}

function createComponent(component: any) {
  DEBUG &&
    console.debug(
      "Analysing component: ",
      component?.name || component?.displayName || component
    );

  return (props: CssProps | undefined, children: ReactNode | ReactNode[]) =>
    createElement(component, props, children);
}

export const withCreate = (component: string | ComponentType): Creatable => {
  if (typeof component === "string") {
    return { create: createComponent(component) } as unknown as Creatable;
  }
  (component as Creatable).create = createComponent(component);
  return component as Creatable;
};

export { withCreate as default, createElement as h };
