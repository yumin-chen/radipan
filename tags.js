export * from './dist/html-tags';
