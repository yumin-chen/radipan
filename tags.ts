export * from "./dist/html-tags";
export type * from "./dist/html-tags.d.ts";
