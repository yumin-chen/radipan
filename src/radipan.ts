import { h } from 'phy-react';
import { FunctionComponent, ReactNode } from 'react';
import { css, cva, cx } from 'radipan/design-system';
import { RecipeProps, CssProps, Creatable } from './radipan.d';
import {
  RecipeDefinition,
  RecipeVariantRecord,
} from '@radipan-design-system/types/recipe';
import { SystemStyleObject } from '@radipan-design-system/types';

const getVariantProps = (props: RecipeProps) => {
  const { css: cssProp, ...restProps } = props;
  return Object.keys(cssProp?.variants || {}).reduce(
    (prev, variantName) => ({
      ...prev,
      [variantName]: restProps?.[variantName as keyof typeof restProps],
    }),
    {}
  );
};

export const parseCssProp = (props: CssProps) => {
  const { css: cssProp } = props;
  // Recipes
  const isRecipe = Object.hasOwn(cssProp || {}, 'variants');
  if (isRecipe) {
    const variantProps = getVariantProps(props);
    return cva(cssProp as RecipeDefinition<RecipeVariantRecord>)(variantProps);
  } else {
    return css(cssProp as SystemStyleObject);
  }
};

export function jsx(component, props, ...children) {
  if (typeof props?.css === 'object') {
    const { css: cssProp, className, ...restProps } = props;
    const cssClasses = parseCssProp(props);
    // const variantProps = typeof cssProp?.variants === 'object' && getVariantProps(props) || null;
    Object.keys(cssProp?.variants || []).forEach(
      variantName => delete restProps[variantName]
    );

    return h(
      component,
      {
        ...restProps,
        // Merge class names with generated styles
        className: !className ? cssClasses : cx(cssClasses, className),
      },
      children
    );
  }

  return children === undefined
    ? h(component, props)
    : h(component, props, children);
};

const createComponent = (component: any) => {
  return (props: any, children: any) => jsx(component, props, children);
};

export const withCreate = (component: any): Creatable => {
  if (typeof component === 'string') {
    return { create: createComponent(component) } as unknown as Creatable;
  }
  (component as Creatable).create = createComponent(component);
  return component as Creatable;
};

export default withCreate;
