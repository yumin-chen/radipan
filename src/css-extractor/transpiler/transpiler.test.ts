import {
  transform,
  transformJsx,
  transformHyperScript,
  transformRadipanSyntax,
  transpile,
} from "./transpiler";
import {
  describe,
  beforeEach,
  afterEach,
  expect,
  jest,
  it,
} from "@jest/globals";
import { readFileSync } from "fs";
import mock from "mock-fs";

jest.mock("../../cli/get-config", () => ({
  __esModule: true,
  outdir: "test_out_dir",
}));

describe("transpiler", () => {
  describe("transform", () => {
    it("should return the transformed code mixed with all styles", () => {
      const src = `
    import React from "react";
    import { p } from "radipan/tags";
    
    function App() {
      return (
        <div>
          <h1 radipanId="uniq-test">Hello World</h1>
          { paragraph }
          { button }
        </div>
      );
    };
    
    const paragraph = p.create({
      anotherProp: "heavy",
      radipanId: "another-uniq-id",
    }, "This is a paragraph");

    const button = h("button", { radipanId: "btn" });
`;

      // Call the function with the sample inputs
      const output1 = transform(src, "uniq-test", "specific-class");
      const output2 = transform(
        output1,
        "another-uniq-id",
        "another-classes and-more"
      );
      const output3 = transform(output2, "btn", "more styling classes");

      // Expect the output to match the expected output
      expect(output1).toMatchSnapshot();
      expect(output2).toMatchSnapshot();
      expect(output3).toMatchSnapshot();
    });
  });

  describe("transformJsx", () => {
    it("should find the identified radipanId from a nested JSX element and transform it", () => {
      const input = `
      <div radipanId="2-def">
        <h1 radipanId="3-ghi">Title</h1>
        <p radipanId="4-jkl">Content</p>
      </div>
    `;
      const output = `<div radipanId="2-def">
        <h1 className="some-class">Title</h1>
        <p radipanId="4-jkl">Content</p>
      </div>;`;
      expect(transformJsx(input, "3-ghi", "some-class")).toBe(output);
    });

    it("should remove any css prop in the element when found the radipanId match", () => {
      const input = `
      <div css={{ color: "red"}} radipanId="asdf">Test</div>
    `;
      const output = `<div className="new-classes">Test</div>;`;
      expect(transformJsx(input, "asdf", "new-classes")).toBe(output);
    });

    it("should remove any prop that is in the keys of the variantProps parameter", () => {
      const input = `
      <div variant="solid" radipanId="asdf">Test</div>
    `;
      const output = `<div className="another-class">Test</div>;`;
      expect(
        transformJsx(input, "asdf", "another-class", { variant: "solid" })
      ).toBe(output);
    });

    it("should return the source code directly if no match of radipanId is found", () => {
      const input = `
      <div variant="solid" radipanId="asdf">Test</div>
    `;
      expect(transformJsx(input, "zxcv", "new-class")).toBe(input);
    });
  });

  describe("transformHyperScript", () => {
    it("should find the identified radipanId from a nested HyperScript call and transform it", () => {
      const input = `
      h("div", { radipanId: "2-def" }, [
        h("h1", { radipanId: "3-ghi" }, "Title"),
        h("p", { radipanId: "4-jkl", css: { color: "red" } }, "Content"),
      ])
    `;

      const output1 = transformHyperScript(input, "3-ghi", "qwerqwer");
      const output2 = transformHyperScript(
        output1,
        "4-jkl",
        "another classes string"
      );
      const output3 = transformHyperScript(output2, "2-def", "poiupoiu");

      expect(output1).toMatchSnapshot();
      expect(output2).toMatchSnapshot();
      expect(output3).toMatchSnapshot();
    });

    it("should remove any css prop in the object when found the radipanId match", () => {
      const input = `
      h("div", { css: { color: "red" }, radipanId: "asdf" }, "Test")
    `;
      const output = `h("div", {
  className: "new-class"
}, "Test");`;
      expect(transformHyperScript(input, "asdf", "new-class")).toBe(output);
    });

    it("should remove any prop that is in the keys of the variantProps parameter", () => {
      const input = `
      h("div", { variant: "solid", radipanId: "zxcv" }, "Test")
    `;
      const output = `h("div", {
  className: "class names"
}, "Test");`;
      expect(
        transformHyperScript(input, "zxcv", "class names", { variant: "solid" })
      ).toBe(output);
    });

    it("should return the source code directly if no match of radipanId is found, even if the props object is ommited", () => {
      const input = `h("div", [h(App)])`;
      expect(transformHyperScript(input, "asdf", "new-class")).toBe(input);
    });
  });

  describe("transformRadipanSyntax", () => {
    it("should transform a Radipan Syntax element with props and children", () => {
      const source = `
      import { div, main } from "radipan/tags";
      function App() {
        return main.create({
          css: {
            width: "100%",
            height: "100vh",
            color: { base: "black" },
            background: { base: "white" }
          },
          radipanId: "11-UF7JEX2"
        }, div.create({
          className: "fs_2xl font_bold"
        }, "Hello Radipüêºn!"));
      }`;
      const radipanId = "11-UF7JEX2";
      const className = "green-div";
      expect(
        transformRadipanSyntax(source, radipanId, className)
      ).toMatchSnapshot();
    });

    it("should add className prop to a Radipan Syntax element with props and children", () => {
      const source =
        'div.create({ css: { color: "green" }, radipanId: "00-XXXXXXX" }, ["Hello"])';
      const radipanId = "00-XXXXXXX";
      const className = "green-div";
      const output = `div.create({
  className: "green-div"
}, ["Hello"]);`;
      expect(transformRadipanSyntax(source, radipanId, className)).toBe(output);
    });

    it("should add className prop to a Radipan Syntax element with props only", () => {
      const source =
        'input.create({ css: { color: "blue" }, placeholder: "blue input", radipanId: "01-XXXXXXX" })';
      const radipanId = "01-XXXXXXX";
      const className = "blue-input";
      const output = `input.create({
  placeholder: "blue input",
  className: "blue-input"
});`;
      expect(transformRadipanSyntax(source, radipanId, className)).toBe(output);
    });

    it("should not add className prop to a Radipan Syntax element with children only", () => {
      const source = 'span.create("Hello")';
      const radipanId = "";
      const className = "hello-span";
      expect(transformRadipanSyntax(source, radipanId, className)).toBe(source);
    });

    it("should not add className prop to a Radipan Syntax element with no props and no children", () => {
      const source = "button.create()";
      const radipanId = "";
      const className = "empty-button";
      const output = "button.create()";
      expect(transformRadipanSyntax(source, radipanId, className)).toBe(output);
    });

    it("should return the source code directly if no match is found", () => {
      const source =
        'p.create({ css: { color: "red" }, radipanId: "02-XXXXXXX" }, ["Hello"])';
      const radipanId = "03-XXXXXXX";
      const className = "red-paragraph";
      const output = source;
      expect(transformRadipanSyntax(source, radipanId, className)).toBe(output);
    });
  });

  describe("transpile", () => {
    let srcFilename = "";
    let srcContent = "";
    const EXPORT_FOLDER = "node_modules/test_out_dir/exported";
    const filenames = {
      jsxFile: "jsxFile.jsx",
      hsFile: "hsFile.ts",
    };

    afterEach(() => {
      mock.restore();
    });

    it("should transpile JSX syntax", async () => {
      const jsxSrc = `
      <div radipanId="radipanId1" css={{ color: "red" }}>
        Hello
      </div>
      `;
      const jsxOutput = `<div className="class1">
        Hello
      </div>;`;
      srcFilename = filenames.jsxFile;
      srcContent = jsxSrc;
      process.env.CSSGEN_FILE = srcFilename;
      mock({
        [`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`]: srcContent,
      });
      // Call the transpile function with some arguments
      const result = await transpile("radipanId1", "class1");
      // Expect the result to be true
      expect(result).toBe(true);
      // Expect the output file to match the expected output
      expect(
        readFileSync(`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`, "utf-8")
      ).toEqual(jsxOutput);
    });

    it("should transpile Hyperscript syntax", async () => {
      const hyperscriptSrc = `
    h("div", { radipanId: "radipanId2", css: { color: "blue" } }, "World")
    `;
      const hyperscriptOutput = `h("div", {
  className: "class2"
}, "World");`;
      srcFilename = filenames.hsFile;
      srcContent = hyperscriptSrc;
      process.env.CSSGEN_FILE = srcFilename;
      mock({
        [`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`]: srcContent,
      });
      // Call the transpile function with some arguments
      const result = await transpile("radipanId2", "class2");
      // Expect the result to be true
      expect(result).toBe(true);
      // Expect the output file to match the expected output
      expect(
        readFileSync(`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`, "utf-8")
      ).toEqual(hyperscriptOutput);
    });

    it("should transpile Radipan syntax", async () => {
      const radipanSrc = `import { div, main } from "radipan/tags";
      function App() {
        return main.create({
          css: {
            width: "100%",
            height: "100vh",
            color: {base: "black",_osDark: "white"},
            background: {base: "white",_osDark: "black"}
          },
          radipanId: "11-UF7JEX2"
        }, div.create({className: "fs_2xl font_bold"}, "Hello Radipüêºn!"));
      }`;

      srcFilename = "another_file.ts";
      srcContent = radipanSrc;
      process.env.CSSGEN_FILE = srcFilename;
      mock({
        [`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`]: srcContent,
      });
      // Call the transpile function with some arguments
      const result = await transpile("11-UF7JEX2", "compiled classes");
      // Expect the result to be true
      expect(result).toBe(true);
      // Expect the output file to match the expected output
      expect(
        readFileSync(`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`, "utf-8")
      ).toMatchSnapshot();
    });

    it("should return false if radipanId is not found", async () => {
      const hyperscriptSrc = `
    h("div", { radipanId: "radipanId2", css: { color: "blue" } }, "World")
    `;
      srcFilename = filenames.hsFile;
      srcContent = hyperscriptSrc;
      process.env.CSSGEN_FILE = srcFilename;
      mock({
        [`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`]: srcContent,
      });
      // Call the transpile function with an invalid radipanId
      const result = await transpile("radipanId3", "class1", {});
      // Expect the result to be false
      expect(result).toBe(false);
    });

    it("should return false if closing tag is not found", async () => {
      srcFilename = filenames.jsxFile;
      srcContent = `<div radipanId={"radipanId4"} css={{ color: "green" }
      Oops
    </div>
    `;
      process.env.CSSGEN_FILE = srcFilename;
      mock({
        [`${EXPORT_FOLDER}/${srcFilename}.lite.tsx`]: srcContent,
      });
      // Call the transpile function with a valid radipanId but an invalid source file
      const result = await transpile("radipanId4", "class1");
      // Expect the result to be false
      expect(result).toBe(false);
    });
  });
});
