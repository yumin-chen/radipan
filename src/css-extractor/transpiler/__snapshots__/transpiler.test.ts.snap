// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transpiler transform should return the transformed code mixed with all styles 1`] = `
"import React from "react";
import { p } from "radipan/tags";
function App() {
  return <div>
          <h1 className="specific-class">Hello World</h1>
          {paragraph}
          {button}
        </div>;
}
;
const paragraph = p.create({
  anotherProp: "heavy",
  radipanId: "another-uniq-id"
}, "This is a paragraph");
const button = h("button", {
  radipanId: "btn"
});"
`;

exports[`transpiler transform should return the transformed code mixed with all styles 2`] = `
"import React from "react";
import { p } from "radipan/tags";
function App() {
  return <div>
          <h1 className="specific-class">Hello World</h1>
          {paragraph}
          {button}
        </div>;
}
;
const paragraph = p.create({
  anotherProp: "heavy",
  className: "another-classes and-more"
}, "This is a paragraph");
const button = h("button", {
  radipanId: "btn"
});"
`;

exports[`transpiler transform should return the transformed code mixed with all styles 3`] = `
"import React from "react";
import { p } from "radipan/tags";
function App() {
  return <div>
          <h1 className="specific-class">Hello World</h1>
          {paragraph}
          {button}
        </div>;
}
;
const paragraph = p.create({
  anotherProp: "heavy",
  className: "another-classes and-more"
}, "This is a paragraph");
const button = h("button", {
  className: "more styling classes"
});"
`;

exports[`transpiler transformHyperScript should find the identified radipanId from a nested HyperScript call and transform it 1`] = `
"h("div", {
  radipanId: "2-def"
}, [h("h1", {
  className: "qwerqwer"
}, "Title"), h("p", {
  radipanId: "4-jkl",
  css: {
    color: "red"
  }
}, "Content")]);"
`;

exports[`transpiler transformHyperScript should find the identified radipanId from a nested HyperScript call and transform it 2`] = `
"h("div", {
  radipanId: "2-def"
}, [h("h1", {
  className: "qwerqwer"
}, "Title"), h("p", {
  className: "another classes string"
}, "Content")]);"
`;

exports[`transpiler transformHyperScript should find the identified radipanId from a nested HyperScript call and transform it 3`] = `
"h("div", {
  className: "poiupoiu"
}, [h("h1", {
  className: "qwerqwer"
}, "Title"), h("p", {
  className: "another classes string"
}, "Content")]);"
`;

exports[`transpiler transformRadipanSyntax should transform a Radipan Syntax element with props and children 1`] = `
"import { div, main } from "radipan/tags";
function App() {
  return main.create({
    className: "green-div"
  }, div.create({
    className: "fs_2xl font_bold"
  }, "Hello Radipüêºn!"));
}"
`;

exports[`transpiler transformRadipanSyntax should transform a comlex Radipan Syntax element with props and children 1`] = `
"import { div, main } from "radipan/tags";
function App() {
  return HeadingBase.create({
    id,
    ...props
  }, [children, !!id && [anchor.create({
    title: 'Direct link to heading',
    onClick: async () => {
      !!copy && (await copy("Test"));
    },
    className: "result of compiled anchor styles"
  })]]);
}"
`;

exports[`transpiler transpile should transpile Radipan syntax 1`] = `
"import { withCreate } from "radipan";
import { div, main } from "radipan/tags";
function App() {
  return main.create({
    className: "compiled classes"
  }, [withCreate("div").create({
    className: "fs_2xl font_bold"
  }, "Hello Radipüêºn!")]);
}"
`;
